src/Game/Chess/Internal.hs

```diff
@@ -393,17 +444,64 @@ doPly p m
 -- can be applied to the position.  This is useful if the move has been generated
 -- by the 'legalPlies' function.
 unsafeDoPly :: Position -> Ply -> Position
-unsafeDoPly pos@Position{color, halfMoveClock, moveNumber} m =
-  (unsafeDoPly' pos m)
-  { color = opponent color
+unsafeDoPly pos@Position{color, halfMoveClock, moveNumber, key, repetitionInfo} m =
+  let pos' = unsafeDoPly' pos m
+  in pos' { color = opponent color
   , halfMoveClock = if isCapture pos m || isPawnPush pos m
                     then 0
                     else halfMoveClock + 1
   , moveNumber = if color == Black
                  then moveNumber + 1
                  else moveNumber
+  , key = key `xor` getPlyHash pos pos'
+  , repetitionInfo = if isCapture pos m || isPawnPush pos m
+      then []
+      -- TODO count how many times this position's hash has occurred before
+      else let ri = RepetitionInfo {
+        rKey = key,
+        rCount = 0
+      } in ri:repetitionInfo
   }
 
+-- | Calculates the difference in board state across a ply, by comparing the pre and post qbb/flags state.
+-- Then returns the Zobrist hash of that difference.
+getPlyHash :: Position -> Position -> Word64
+getPlyHash pos@Position{qbb, flags} pos'@Position{qbb=qbb', flags=flags'} =
```

Comment:
Strictness?

Reply:
Here I was wondering if I should use bang patterns in the Position arguments, but followed the pattern from unsafeDoPly and omitted them.